osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/test/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/firmware/UpdateFirmwareCommandExecutorIntegrationTest.java

```diff
@@ -99,12 +98,77 @@ void testExecute() throws Exception {
+        RandomStringUtils.randomAlphabetic(100).getBytes(StandardCharsets.UTF_8);
+
+    final String firmwareImageIdentifier = "496d6167654964656e746966696572";
+
     when(this.firmwareFileCachingRepository.retrieve(firmwareIdentification))
         .thenReturn(firmwareFile);
+    when(this.firmwareImageIdentifierCachingRepository.retrieve(firmwareIdentification))
+        .thenReturn(firmwareImageIdentifier);
 
+    final UpdateFirmwareRequestDto updateFirmwareRequestDto =
+        new UpdateFirmwareRequestDto(firmwareIdentification, deviceIdentification);
     this.commandExecutor.execute(
-        this.connectionManagerStub, device, firmwareIdentification, this.messageMetadata);
+        this.connectionManagerStub, device, updateFirmwareRequestDto, this.messageMetadata);
+
+    verify(this.dlmsDeviceRepository, never()).findByDeviceIdentification(deviceIdentification);
+    verify(this.macGenerationService, never()).calculateMac(any(), any(), any());
+    verify(this.firmwareFileCachingRepository, times(1)).retrieve(firmwareIdentification);
+    verify(this.firmwareImageIdentifierCachingRepository, times(1))
+        .retrieve(firmwareIdentification);
+
+    this.doAsserts();
+  }
 
+  @Test
+  void testExecuteMbusFirmware() throws Exception {
+    final DlmsDevice device = new DlmsDevice();
+    device.setMbusIdentificationNumber(1L);
+    final String firmwareIdentification = RandomStringUtils.randomAlphabetic(10);
+    final String deviceIdentification = RandomStringUtils.randomAlphabetic(10);
+    device.setDeviceIdentification(deviceIdentification);
+
+    final byte[] firmwareFile =
+        org.bouncycastle.util.encoders.Hex.decode(
+            "534d523500230011004000310000001000020801e91effffffff500303000000000000831c9d5aa5b4f"
+                + "fbfd057035a8a7896a4abe7afa36687fbc48944bcee0343eed3a75aab882ec1cf57820adfd4394e262d"
+                + "5fa821c678e71c05c47e1c69c4bfffe1fd");
+
+    when(this.dlmsDeviceRepository.findByDeviceIdentification(deviceIdentification))
+        .thenReturn(device);
+    when(this.firmwareFileCachingRepository.retrieve(firmwareIdentification))
+        .thenReturn(firmwareFile);
+    when(this.macGenerationService.calculateMac(any(), any(), any())).thenReturn(new byte[16]);
+
+    final UpdateFirmwareRequestDto updateFirmwareRequestDto =
+        new UpdateFirmwareRequestDto(firmwareIdentification, deviceIdentification);
+    this.commandExecutor.execute(
+        this.connectionManagerStub, device, updateFirmwareRequestDto, this.messageMetadata);
+
+    verify(this.dlmsDeviceRepository, times(1)).findByDeviceIdentification(deviceIdentification);
+    verify(this.macGenerationService, times(1)).calculateMac(any(), any(), any());
+    verify(this.firmwareFileCachingRepository, times(1)).retrieve(firmwareIdentification);
+    verify(this.firmwareImageIdentifierCachingRepository, never()).retrieve(firmwareIdentification);
+
+    this.doAsserts();
+  }
+
+  private void doAsserts() {
```

Comment:
Perhaps the tests this is called from would read a bit nicer if this had a name like "assertImageTransferRelatedInteractionWithConnection"? (Some specific info instead of the general "asserts" that says nothing about what is asserted.)

Reply:
renamed method